### PR TITLE
Remove label names from Zinc for records and tuples

### PR DESCRIPTION
**Pull request by Zett98**
## Depends

- [x] #321

## Problem

Records are represented as Stack_item.t LMap.t, which is a Map underneath that uses labels (which are strings) as keys
## Solution

Turn labels used by Records into ints and Map to Array to allow simple o(1) access to the field value
- #163

### CODE REVIEW

File: ligolang/src/passes/13deku-zincing/compiler.ml [hunk 1]
```diff
@@ -170,24 +170,34 @@ and other_compile :
   (* Variant *)
   | E_constructor { constructor = Label constructor; element } ->
       compile_known_function_application environment
-        (fun ~k -> MakeVariant (Zinc_utils.Label constructor) :: k)
+        (fun ~k -> MakeVariant ( constructor) :: k)
```
Comment (rosalogia) [inline ligolang/src/passes/13deku-zincing/compiler.ml:173]: Formatting

File: ligolang/src/passes/13deku-zincing/compiler.ml [hunk 2]
```diff
@@ -209,9 +219,9 @@ and compile_constant :
   | C_FAILWITH -> Failwith :: k
   | C_CONTRACT_OPT -> Contract_opt :: k
   | C_CALL -> MakeTransaction :: k
-  | C_UNIT -> MakeRecord [] :: k
-  | C_NONE -> MakeRecord [] :: MakeVariant (Label "None") :: k
-  | C_SOME -> MakeVariant (Label "Some") :: k
+  | C_UNIT -> MakeRecord 0 :: k
+  | C_NONE -> MakeRecord 0 :: MakeVariant ("None") :: k
+  | C_SOME -> MakeVariant ( "Some") :: k
```
Comment (rosalogia) [inline ligolang/src/passes/13deku-zincing/compiler.ml:224]: Paren space again :smile:

Comment (rosalogia) [reply ligolang/src/passes/13deku-zincing/compiler.ml:224]: Hm, are you trying to align with something?

File: zinc/lib/Zinc_types.ml [hunk 3]
```diff
@@ -35,10 +35,10 @@ type 'a zinc_instruction =
   | Num of Z.t
   | Add
   (* ASTs *)
-  | MakeRecord of label list
+  | MakeRecord of label
```
Comment (anchpop) [inline zinc/lib/Zinc_types.ml:38]:
```suggestion
  | MakeRecord of int
```
I know I suggested this but it was actually wrong

File: zinc/lib/Zinc_utils.ml [hunk 4]
```diff
@@ -19,28 +19,38 @@ module Z = struct
...
+type variant_label = string [@@deriving show {with_path = false}, eq, yojson]
 
 module LMap = struct
-  include Map.Make (struct
-    type t = label
+  type 'a t = 'a array [@@deriving yojson]
 
-    let compare = compare
-  end)
+  type 'a json = 'a list [@@deriving yojson]
```
Comment (anchpop) [inline zinc/lib/Zinc_utils.ml:29]:
```suggestion
```
I'm not sure what purpose this serves

File: zinc/lib/Zinc_utils.ml [hunk 5]
```diff
@@ -19,28 +19,38 @@ module Z = struct
...
 
-type label = Label of string [@@deriving show {with_path = false}, eq, yojson]
+type label = int [@@deriving show {with_path = false}, eq, yojson]
+
+type variant_label = string [@@deriving show {with_path = false}, eq, yojson]
 
 module LMap = struct
-  include Map.Make (struct
-    type t = label
+  type 'a t = 'a array [@@deriving yojson]
```
Comment (anchpop) [inline zinc/lib/Zinc_utils.ml:27]:
```suggestion
  type 'a t = 'a array [@@deriving show, eq, yojson]
```
if you add this you don't need to write pp, eq, or json definitions

File: zinc/lib/Zinc_utils.ml [hunk 6]
```diff
@@ -19,28 +19,38 @@ module Z = struct
...
 
-  let of_list (lst : 'a association_list) : 'a t =
-    let aux prev (k, v) = add k v prev in
-    List.fold_left aux empty lst
+  let of_list (lst : 'a list) : 'a t = lst |> Array.of_list
+
+  let equal eq a b =
+    if Int.equal (Array.length a) (Array.length b) then
+      List.equal eq (Array.to_list a) (Array.to_list b)
+    else false
```
Comment (anchpop) [inline zinc/lib/Zinc_utils.ml:38]:
```suggestion
```

File: zinc/lib/Zinc_utils.ml [hunk 7]
```diff
@@ -54,9 +64,10 @@ module LMap = struct
          new_pp)
       lst
 
-  let of_yojson a m = Result.map of_list (association_list_of_yojson a m)
+  let of_yojson a m = Result.map of_list (json_of_yojson a m)
 
-  let to_yojson a m = bindings m |> association_list_to_yojson a
+  let to_yojson a m =
+    Array.to_seqi m |> List.of_seq |> association_list_to_yojson a
```
Comment (anchpop) [inline zinc/lib/Zinc_utils.ml:70]:
```suggestion
```

**Review by anchpop — APPROVED** [review]
I haven't tested it extensively, but it looks great

File: zinc/lib/Zinc_utils.ml (hunk 7 above)
Comment (Zett98) [reply zinc/lib/Zinc_utils.ml]: removed

File: zinc/lib/Zinc_utils.ml (hunk 6 above)
Comment (Zett98) [reply zinc/lib/Zinc_utils.ml]: removed

File: zinc/lib/Zinc_utils.ml (hunk 5 above)
Comment (Zett98) [reply zinc/lib/Zinc_utils.ml]: done

File: zinc/lib/Zinc_utils.ml (hunk 4 above)
Comment (Zett98) [reply zinc/lib/Zinc_utils.ml]: it was a dumb type that wasn't needed

File: zinc/lib/Zinc_types.ml (hunk 3 above)
Comment (Zett98) [reply zinc/lib/Zinc_types.ml]: done

File: ligolang/src/passes/13deku-zincing/compiler.ml (hunk 2 above)
Comment (Zett98) [reply ligolang/src/passes/13deku-zincing/compiler.ml]: changed formatting

File: ligolang/src/passes/13deku-zincing/compiler.ml (hunk 1 above)
Comment (Zett98) [reply ligolang/src/passes/13deku-zincing/compiler.ml]: changed

**Review by anchpop — APPROVED** [review]

**Comment by anchpop** [comment]
@rosalogia Does this PR look okay to you?

**Review by rosalogia — APPROVED** [review]

**Comment by anchpop** [comment]
anime girl saying thank you

**Pull request merged**